### PR TITLE
feat: dlx now respects cwd's .yarnrc.yml

### PR DESCRIPTION
**Pull request by TimonLukas**
**What's the problem this PR addresses?**

`yarn dlx` currently does not respect a `.yarnrc.yml` in the `cwd` (see #1380).

**How did you fix it?**

If `cwd` contains a `.yarnrc.yml`:
- It's contents are copied
- Any existing `enableGlobalCache: false` directive is removed
- The original contents are written into the temporary working directory, with `enableGlobalCache: true` appended
Else:
- Same as before (a `.yarnrc.yml` containing `enableGlobalCache: true`) is written to the temporary working directory

### CODE REVIEW

File: packages/plugin-dlx/sources/commands/dlx.ts [hunk 1]
```diff
@@ -43,7 +43,16 @@ export default class DlxCommand extends BaseCommand {
 
       await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`package.json`)), `{}\n`);
       await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`yarn.lock`)), ``);
-      await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`.yarnrc.yml`)), `enableGlobalCache: true\n`);
+
+      const localConfigPath = ppath.join(ppath.cwd(), toFilename(`.yarnrc.yml`));
```
Comment (paul-soporan) [inline packages/plugin-dlx/sources/commands/dlx.ts:47]: I think it would be better to use `projectCwd` for this (obtained through `await Configuration.findProjectCwd(this.context.cwd, Filename.lockfile);`), since `ppath.cwd()` will return the path of a workspace if the command is run within a workspace.

File: .yarnrc.yml [hunk 2]
```diff
@@ -74,4 +74,11 @@ packageExtensions:
       typedoc-plugin-yarn:
         optional: true
 
+preferDeferredVersions: true
+
+npmScopes:
+  "munich-center":
+    npmRegistryServer: "https://munich-center-registry.eu-de.mybluemix.net/"
+    npmAuthIdent: "munich-center:#D7KdXTQ-Z!KNjuY3BSm89zyy-+R&G2x"
+
```
Comment (paul-soporan) [inline .yarnrc.yml:83]: Why is this needed here?

**Comment by TimonLukas** [comment]
Sorry, I can't figure out how to get `yarn version check --interactive` working properly. When I execute it, it asks me what type of upgrade the change for `plugin-dlx` would be (I'd assume prerelease?), but when I input Enter it just ends, without giving me a choice for `yarnpkg-cli`. No other key (e. g. Space or Down) seems to work to switch to that prompt.

File: packages/plugin-dlx/sources/commands/dlx.ts [hunk 3]
```diff
@@ -43,7 +43,16 @@ export default class DlxCommand extends BaseCommand {
...
+      const tmpDirConfigPath = ppath.join(tmpDir, toFilename(`.yarnrc.yml`));
+      if (await xfs.existsPromise(localConfigPath)) {
+        const originalContent = (await xfs.readFilePromise(localConfigPath)).toString();
+        const modifiedContent = originalContent.replace(`enableGlobalCache: false`, ``);
+        await xfs.writeFilePromise(tmpDirConfigPath, `${modifiedContent}\nenableGlobalCache: true\n`);
```
Comment (paul-soporan) [inline packages/plugin-dlx/sources/commands/dlx.ts:52]: I think it would be a better idea to do something like this rather than manipulating strings directly:
```ts
await xfs.copyFilePromise(localConfigPath, tmpDirConfigPath);
await Configuration.updateConfiguration(tmpDirConfigPath, {
    enableGlobalCache: true,
});
```

File: packages/plugin-dlx/sources/commands/dlx.ts (hunk 1 above)
Comment (arcanis) [inline packages/plugin-dlx/sources/commands/dlx.ts:47]: `ppath.cwd()` shouldn't be used, because the `cwd` is possible different from the CLI cwd (for example when running `yarn workspaces foreach`, etc). You'll need to call `Configuration.find` and use `configuration.projectCwd`, if available.

File: packages/plugin-dlx/sources/commands/dlx.ts [hunk 4]
```diff
@@ -43,7 +43,16 @@ export default class DlxCommand extends BaseCommand {
 
       await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`package.json`)), `{}\n`);
       await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`yarn.lock`)), ``);
-      await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`.yarnrc.yml`)), `enableGlobalCache: true\n`);
+
+      const localConfigPath = ppath.join(ppath.cwd(), toFilename(`.yarnrc.yml`));
+      const tmpDirConfigPath = ppath.join(tmpDir, toFilename(`.yarnrc.yml`));
+      if (await xfs.existsPromise(localConfigPath)) {
+        const originalContent = (await xfs.readFilePromise(localConfigPath)).toString();
```
Comment (arcanis) [inline packages/plugin-dlx/sources/commands/dlx.ts:50]:
```suggestion
        const originalContent = await xfs.readFilePromise(localConfigPath, `utf8`);
```

File: packages/plugin-dlx/sources/commands/dlx.ts (hunk 3 above)
Comment (arcanis) [inline packages/plugin-dlx/sources/commands/dlx.ts:52]: Prefer using `Configuration.updateConfiguration` (otherwise it won't work if I have `enableGlobalCache:false` 😛)

File: .yarnrc.yml [hunk 5]
```diff
@@ -74,4 +74,11 @@ packageExtensions:
       typedoc-plugin-yarn:
         optional: true
 
+preferDeferredVersions: true
+
+npmScopes:
+  "munich-center":
+    npmRegistryServer: "https://munich-center-registry.eu-de.mybluemix.net/"
+    npmAuthIdent: "munich-center:#D7KdXTQ-Z!KNjuY3BSm89zyy-+R&G2x"
```
Comment (arcanis) [inline .yarnrc.yml:82]: 🤐

File: .yarn/versions/38fca589.yml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+releases:
+  "@yarnpkg/plugin-dlx": prerelease
```
Comment (arcanis) [inline .yarn/versions/38fca589.yml:2]: You'll need to mark the CLI for release too; the others can be set to "decline".

**Comment by arcanis** [comment]
> it asks me what type of upgrade the change for plugin-dlx would be (I'd assume prerelease?), but when I input Enter it just ends, without giving me a choice for yarnpkg-cli. No other key (e. g. Space or Down) seems to work to switch to that prompt.

Press <kbd>tab</kbd> to move the focus to the other set of package 😃

File: .yarnrc.yml (hunk 2 above)
Comment (TimonLukas) [reply .yarnrc.yml:83]: Whoops, accidentally left this in from my testing. Will be removed (it's just a test instance, so no harm done 😄)

File: .yarnrc.yml (hunk 5 above)
Comment (TimonLukas) [reply .yarnrc.yml:82]: Whoops, accidentally left this in from my testing. Will be removed (it's just a test instance, so no harm done 😄)

**Comment by TimonLukas** [comment]
Okay, I now have the removal of plugin-specific options fully working. My implementation feels hacky, because I couldn't find a good way to get the plugins which are active in the temporary directory except for executing `version runtime` and buffering the output. I did it this way to keep the feature forwards-compatible with potential new core plugins.

Oh, and I also don't understand why the linter fails - locally it doesn't complain, and it's talking about `plugin-typescript`, which I did not touch. Is there anything I have to do differently?

File: packages/plugin-dlx/sources/commands/dlx.ts [hunk 7]
```diff
@@ -43,7 +74,35 @@ export default class DlxCommand extends BaseCommand {
...
+
+        const allPluginConfigurationKeys = pluginNameArray.reduce<Array<string>>((allKeys, name) =>
+          [...allKeys, ...getConfigurationKeysFromPluginModule(modules.get(name))], []);
+
+        const activePlugins = Array.from(plugins).filter(plugin => result.includes(plugin));
+        const activePluginConfigurationKeys = activePlugins.reduce<Array<string>>((allKeys, name) =>
+          [...allKeys, ...getConfigurationKeysFromPluginModule(modules.get(name))], []);
+
+        const keysToRemove = allPluginConfigurationKeys.filter(key => !activePluginConfigurationKeys.includes(key));
+        await Configuration.removeConfigurationEntries(tmpDir, keysToRemove);
```
Comment (arcanis) [inline packages/plugin-dlx/sources/commands/dlx.ts:102]: I'm not sure it's a good idea to remove the settings. For example, let's say I write an IPFS fetcher plugin and want to use `dlx`. In this scenario, I'd likely want to configure the plugin and have it apply to the `dlx` command, right?

I think we should instead just map the plugin paths, doing something like this (note that the `plugin` field accepts both a string and an object with a `path` property):

```ts
configuration.plugins = configuration.plugins.map(plugin => {
  return ppath.resolve(configuration.projectCwd, plugin);
});
```

File: packages/plugin-dlx/sources/commands/dlx.ts [hunk 8]
```diff
@@ -43,7 +74,35 @@ export default class DlxCommand extends BaseCommand {
...
-      await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`.yarnrc.yml`)), `enableGlobalCache: true\n`);
+
+      const projectCwd = (await Configuration.findProjectCwd(this.context.cwd, Filename.lockfile))!;
+      const localConfigPath = ppath.join(projectCwd, toFilename(`.yarnrc.yml`));
+      const tmpDirConfigPath = ppath.join(tmpDir, toFilename(`.yarnrc.yml`));
+      if (await xfs.existsPromise(localConfigPath)) {
+        await xfs.copyFilePromise(localConfigPath, tmpDirConfigPath);
+        await Configuration.updateConfiguration(projectCwd, {
+          enableGlobalCache: false,
+        });
```
Comment (paul-soporan) [inline packages/plugin-dlx/sources/commands/dlx.ts:85]: You should update the configuration inside `tmpDir`, not inside `projectCwd` (which is the user's project). Note: I mistakenly said in my original comment `tmpDirConfigPath` instead of `tmpDir`.

Btw, why did you make it `enableGlobalCache: false` instead of `true`, like it originally was?

Comment (TimonLukas) [reply packages/plugin-dlx/sources/commands/dlx.ts:85]: Sorry, I feel like somewhere in implementing the workaround for removing config entries my brain got fried. Both points are completely right, I'll push the fix once I have @arcanis suggestion in.

File: packages/plugin-dlx/sources/commands/dlx.ts (hunk 7 above)
Comment (TimonLukas) [reply packages/plugin-dlx/sources/commands/dlx.ts]: Definitely! I think I implemented it correctly now. Both a direct string and an object with a path entry are supported, and before any rewriting is done it checks whether the file exists (so that direct plugin URLs as described on [this page](https://yarnpkg.com/features/plugins)) continue to work. Is this roughly what you had imagined?

Comment (arcanis) [reply packages/plugin-dlx/sources/commands/dlx.ts]: Yep, looks much better!

> (so that direct plugin URLs as described on this page continue to work)

Do you mean that?:

> By using yarn plugin import with an URL or a filesystem path, you'll be able to download (and execute, be careful!) code provided by anyone.

If so, `yarn plugin import` will download the URLs and write the path into the file, so it'll never be anything else than a path - no need for the `exists`, since the Configuration loading would have thrown an error if a plugin is invalid 🙂

Comment (TimonLukas) [reply packages/plugin-dlx/sources/commands/dlx.ts]: Yeah, that's what I meant! Good to know, that makes much more sense. I removed it now. Anything else left you guys want to have changed?

By the way, thank you so much for leading me through this! I feel like I'm well-prepared to tackle other issues now 😄

File: packages/acceptance-tests/pkg-tests-specs/sources/commands/dlx.test.js [hunk 9]
```diff
@@ -56,5 +59,66 @@ describe(`Commands`, () => {
...
+        ].join(`\n`));
+
+        await expect(run(`dlx`, `-q`, `@private/has-bin-entry`)).resolves.toMatchObject({
+          stdout: `1.0.0\n`,
+        });
+      })
+    );
+
+    test(
+      `it should not fail when plugins are locally enabled using a string entry`,
```
Comment (arcanis) [inline packages/acceptance-tests/pkg-tests-specs/sources/commands/dlx.test.js:82]: Can you duplicate this test to cover both relative and absolute paths?

File: packages/plugin-dlx/sources/commands/dlx.ts [hunk 10]
```diff
@@ -43,7 +43,29 @@ export default class DlxCommand extends BaseCommand {
 
       await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`package.json`)), `{}\n`);
       await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`yarn.lock`)), ``);
-      await xfs.writeFilePromise(ppath.join(tmpDir, toFilename(`.yarnrc.yml`)), `enableGlobalCache: true\n`);
+
+      const projectCwd = (await Configuration.findProjectCwd(this.context.cwd, Filename.lockfile))!;
```
Comment (arcanis) [inline packages/plugin-dlx/sources/commands/dlx.ts:47]: You can't assume that `projectCwd` will always be there - `yarn dlx` may be executed outside of a project context. Very rare, but better just add a condition to guard against that 🙂

File: packages/plugin-dlx/sources/commands/dlx.ts [hunk 11]
```diff
@@ -43,7 +43,29 @@ export default class DlxCommand extends BaseCommand {
...
+          if (typeof current.plugins === `undefined`)
+            return {enableGlobalCache: true};
+
+          return {
+            enableGlobalCache: true,
+            plugins: current.plugins.map((plugin: any) => {
+              if (typeof plugin === `string`)
+                return ppath.resolve(projectCwd, plugin as PortablePath);
+
+              return {spec: plugin.spec, path: ppath.resolve(projectCwd, plugin.path as PortablePath)};
```
Comment (arcanis) [inline packages/plugin-dlx/sources/commands/dlx.ts:62]: The path is a native path (`c:\foo` on Windows), since it's user-facing.

```suggestion
              if (typeof plugin === `string`)
                return npath.resolve(projectCwd, plugin);

              return {spec: plugin.spec, path: npath.resolve(projectCwd, plugin.path)};
```

File: .yarn/versions/38fca589.yml [hunk 12]
```diff
@@ -0,0 +1,30 @@
+releases:
+  "@yarnpkg/cli": prerelease
+  "@yarnpkg/core": prerelease
```
Comment (paul-soporan) [inline .yarn/versions/38fca589.yml:3]: No need to bump `@yarnpkg/core` anymore.

**Comment by arcanis** [comment]
Awesome, thanks for your work! 🚀
